From fb3699633289e51896f3bd302b9e93ebedf8936a Mon Sep 17 00:00:00 2001
From: cnauroth <cnauroth@apache.org>
Date: Tue, 22 Dec 2015 10:29:35 -0800
Subject: [PATCH 1091/2848] HDFS-9458. TestBackupNode always binds to port
 50070, which can cause bind failures. Contributed
 by Xiao Chen.

(cherry picked from commit 5c0ff69618fa303078c0f9afef4c93a864ab6cb7)
(cherry picked from commit 9b039a5004685d61375a7c0d40de95e4c15ae1d6)

Change-Id: I3dfa74990b47a7bf122e323e93e05671e6e42d7b
---
 .../hdfs/server/namenode/TestBackupNode.java       |    9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestBackupNode.java b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestBackupNode.java
index 354248b..30f18d1 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestBackupNode.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestBackupNode.java
@@ -49,6 +49,7 @@
 import org.apache.hadoop.hdfs.server.namenode.FileJournalManager.EditLogFile;
 import org.apache.hadoop.hdfs.server.protocol.NamenodeProtocols;
 import org.apache.hadoop.net.NetUtils;
+import org.apache.hadoop.net.ServerSocketUtil;
 import org.apache.hadoop.security.SecurityUtil;
 import org.apache.hadoop.security.UserGroupInformation;
 import org.apache.hadoop.test.GenericTestUtils;
@@ -144,8 +145,9 @@ public void startBackupNodeWithIncorrectAuthentication() throws IOException {
     Configuration c = new HdfsConfiguration();
     StartupOption startupOpt = StartupOption.CHECKPOINT;
     String dirs = getBackupNodeDir(startupOpt, 1);
-    c.set(DFSConfigKeys.FS_DEFAULT_NAME_KEY, "hdfs://127.0.0.1:1234");
-    c.set(DFSConfigKeys.DFS_NAMENODE_BACKUP_HTTP_ADDRESS_KEY, "localhost:0");
+    c.set(DFSConfigKeys.FS_DEFAULT_NAME_KEY,
+        "hdfs://127.0.0.1:" + ServerSocketUtil.getPort(0, 100));
+    c.set(DFSConfigKeys.DFS_NAMENODE_HTTP_ADDRESS_KEY, "127.0.0.1:0");
     c.set(DFSConfigKeys.DFS_BLOCKREPORT_INITIAL_DELAY_KEY, "0");
     c.setInt(DFSConfigKeys.DFS_DATANODE_SCAN_PERIOD_HOURS_KEY,
         -1); // disable block scanner
@@ -180,7 +182,8 @@ public void startBackupNodeWithIncorrectAuthentication() throws IOException {
           bn.getNamesystem() == null);
       fail("Incorrect authentication setting should throw IOException");
     } catch (IOException e) {
-      LOG.info("IOException thrown as expected", e);
+      LOG.info("IOException thrown.", e);
+      assertTrue(e.getMessage().contains("Running in secure mode"));
     } finally {
       if (nn != null) {
         nn.stop();
-- 
1.7.9.5

